### ISSUE TITLE
Possible exceptions are not documented

### ISSUE THREAD

**Issue by benureau** [sciunto-org/python-bibtexparser#212]
On a parsing problem, most of the time, the user will receive either a `pyparsing.ParseException` or a `pyparsing.ParseSyntaxException`. But when providing the empty string, an `IndexError` will be raised. This is not a problem per se, but currently, this is not documented, and must be discovered when problem happen.

**Comment by benureau** [comment]
And the following bibstring will throw a `bibtexparser.bibdatabase.UndefinedString`:
```
@misc{this looks,
  like = a = bibtex file but
    , is not a real one!
```

```python
bibraw = '''@misc{this looks,
  like = a = bibtex file but
    , is not a real one!
'''

from bibtexparser.bparser import BibTexParser

btp = BibTexParser()

print(btp.parse(bibraw, partial=False).get_entry_dict())
```
produces:

```
Traceback (most recent call last):
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/bibtexparser/bibdatabase.py", line 102, in expand_string
    self.strings[name])
KeyError: 'a'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "pubs/test.py", line 23, in <module>
    print(btp.parse(bibraw, partial=False).get_entry_dict())
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/bibtexparser/bparser.py", line 147, in parse
    self._expr.parseFile(bibtex_file_obj)
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/bibtexparser/bibtexexpression.py", line 263, in parseFile
    return self.main_expression.parseFile(file_obj, parseAll=True)
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 2163, in parseFile
    return self.parseString(file_contents, parseAll)
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 1622, in parseString
    loc, tokens = self._parse( instring, 0 )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 1379, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 3919, in parseImpl
    return super(ZeroOrMore, self).parseImpl(instring, loc, doActions)
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 3848, in parseImpl
    loc, tokens = self_expr_parse( instring, loc, doActions, callPreParse=False )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 1379, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 3530, in parseImpl
    ret = e._parse( instring, loc, doActions )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 1379, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 3395, in parseImpl
    loc, exprtokens = e._parse( instring, loc, doActions )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 1379, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 3530, in parseImpl
    ret = e._parse( instring, loc, doActions )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 1379, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 3395, in parseImpl
    loc, exprtokens = e._parse( instring, loc, doActions )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 1379, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 3378, in parseImpl
    loc, resultlist = self.exprs[0]._parse( instring, loc, doActions, callPreParse=False )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 1379, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 3717, in parseImpl
    return self.expr._parse( instring, loc, doActions, callPreParse=False )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 1379, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 3395, in parseImpl
    loc, exprtokens = e._parse( instring, loc, doActions )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 1379, in _parseNoCache
    loc,tokens = self.parseImpl( instring, preloc, doActions )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 3530, in parseImpl
    ret = e._parse( instring, loc, doActions )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 1405, in _parseNoCache
    tokens = fn( instring, tokensStart, retTokens )
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/pyparsing.py", line 1049, in wrapper
    ret = func(*args[limit[0]:])
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/bibtexparser/bibtexexpression.py", line 260, in _string_expr_parse_action
    return self._string_expr_parse_action_fun(s, l, t)
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/bibtexparser/bparser.py", line 184, in <lambda>
    BibDataStringExpression.expression_if_needed(t)))
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/bibtexparser/bparser.py", line 178, in <lambda>
    maybe_interpolate = lambda expr: as_text(expr)
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/bibtexparser/bibdatabase.py", line 223, in as_text
    return text_string_or_expression.get_value()
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/bibtexparser/bibdatabase.py", line 184, in get_value
    return ''.join([BibDataString.expand_string(s) for s in self.expr])
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/bibtexparser/bibdatabase.py", line 184, in <listcomp>
    return ''.join([BibDataString.expand_string(s) for s in self.expr])
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/bibtexparser/bibdatabase.py", line 150, in expand_string
    return string_or_bibdatastring.get_value()
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/bibtexparser/bibdatabase.py", line 131, in get_value
    return self._bibdatabase.expand_string(self.name)
  File "/Users/fcyb/.venvs/pubs37/lib/python3.7/site-packages/bibtexparser/bibdatabase.py", line 104, in expand_string
    raise(UndefinedString(name))
bibtexparser.bibdatabase.UndefinedString: 'a'
```

**Comment by omangin** [comment]
The empty string part is trivial to fix (see #214).

Actually the second example currently does not raise any exception as mentioned in #211.

Nevertheless this is indeed a separate issue. String checking should probably occur at a higher level of the parsing (as a validation of the full value for example) since here it is tried while pyparsing is still deciding whether the current chunk is indeed a proper `key = value,` definition.
